### PR TITLE
draft for dhcp server testcase

### PR DESCRIPTION
**Pull request by wen587**
### Description of PR

Summary: Testcase of dhcp server for generic updater apply-patch
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
[End to End test support for Generic Updater apply-patch](https://github.com/ghooo/SONiC/blob/c1f3f3b5427d0cafb3defd93df8b906a26fcee8a/doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md#91-unit-tests-for-apply-patch)

### CODE REVIEW

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    '''
+    cmds = []
+
+    # Generate 4 dhcp servers for each new created vlan
+    for vlan in vlan_intfs_list:
+        for i in range(1, 5):
+            cmds.append('config vlan dhcp_relay add {} 192.0.{}.{}'.format(vlan['vlan_id'], vlan['vlan_id'], i))
+    
+    duthost.shell_cmds(cmds=cmds)
+    time.sleep(5)
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_dhcp_relay.py:82]: >time.sleep(5) [](http://example.com/codeflow?start=4&length=13)

We try not to sleep if possible.
If impossible, could you comment that what you are waiting for?

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    pytest_assert(duthost.is_service_fully_started('dhcp_relay'), "dhcp_relay service is not running during setup")
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_dhcp_relay.py:83]: >pytest_assert [](http://example.com/codeflow?start=4&length=13)

Could you also verify that the config takes effect by checking the command line parameters of `dhcpmon` ?

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    for vlan in vlan_intfs_list:
+        cmds.append('config vlan add {}'.format(vlan['vlan_id']))
+        cmds.append("config interface ip add Vlan{} {}".format(vlan['vlan_id'], vlan['ip'].upper()))
+        cmds.append('config vlan member add {} {}'.format(vlan['vlan_id'], rand_vlan_port))
+    
+    logger.info("Commands: {}".format(cmds))
+    duthost.shell_cmds(cmds=cmds)
+
+def create_test_dhcp_servers(duthost, vlan_intfs_list):
+    '''Generate 4 dhcp server for each vlan
```
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply tests/generic_config_updater/test_dhcp_relay.py:57]: Also design for DHCPv6? It is a relative new feature, but we can still *design* for it.

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+    logger.info("Commands: {}".format(cmds))
+    output = duthost.shell(cmds)
+
+    pytest_assert("Patch applied successfully" in output['stdout'], "apply-patch is not working correctly")
+
+DUT_ADD_SERVERS_JSON_FILE='/tmp/dhcp_add_servers.json'
+DUT_DEL_SERVERS_JSON_FILE='/tmp/dhcp_del_servers.json'
+def test_dhcp_relay_tc2_servers(duthost, setup_vlan):
+    """Test add and del dhcp servers cases
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_dhcp_relay.py:150]: >add [](http://example.com/codeflow?start=12&length=3)

There is some corner cases for "add":
1. reach maximal servers
2. add an existing one

File: tests/vlan/test_vlan.py [hunk 5]
```diff
@@ -29,56 +28,24 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
...
-        for addr in cfg_facts['VLAN_INTERFACE']['Vlan'+vlanid]:
-            if addr.find(':') == -1:
-                ip = addr
-                break
-        else:
-            continue
-        logger.info("Original VLAN {}, ip {}".format(vlanid, ip))
-        vlan_intfs_dict[int(vlanid)] = {'ip': ip, 'orig': True}
+def vlan_intfs_dict(tbinfo, utils_vlan_intfs_dict_orig):
+    vlan_intfs_dict = utils_vlan_intfs_dict_orig
```
[the remark this answers is not in the document]
Comment (wen587) [reply tests/vlan/test_vlan.py:33]: Yes. It works.

File: tests/generic_config_updater/dhcp_utils.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    return output
+
+def expect_op_success(duthost, output):
+    pytest_assert(not output['rc'], "Command is not running successfully")
+    pytest_assert("Patch applied successfully" in output['stdout'], "Please check if json file is validate")
+    time.sleep(5)
+    pytest_assert(duthost.is_service_fully_started('dhcp_relay'), "dhcp_relay service is not running")
+
+def expect_res_success(duthost, expected_server_list, unexpected_server_list):
+    output = duthost.shell('docker exec dhcp_relay ps aux')
```
[the remark this answers is not in the document]
Comment (wen587) [reply tests/generic_config_updater/dhcp_utils.py:26]: I have not found a way to only check the process. Each VLAN will have its own process. It looks like this:
root          37  0.0  0.1   9668  3152 pts/0    S    Oct29   0:00 /usr/sbin/dhcrelay -d -m discard -a %h:%p %P --name-alias-map-file /tmp/port-name-alias-map.txt -id Vlan100 -iu Vlan1000 -iu PortChannel0001 -iu PortChannel0002 -iu PortChannel0003 -iu PortChannel0004 192.0.100.1
root          38  0.0  0.1   9668  3160 pts/0    S    Oct29   0:00 /usr/sbin/dhcrelay -d -m discard -a %h:%p %P --name-alias-map-file /tmp/port-name-alias-map.txt -id Vlan1000 -iu Vlan100 -iu PortChannel0001 -iu PortChannel0002 -iu PortChannel0003 -iu PortChannel0004 192.0.0.1 192.0.0.2 192.0.0.3 192.0.0.4

File: tests/generic_config_updater/test_dhcp_relay.py (hunk 1 above)
Comment (wen587) [reply tests/generic_config_updater/test_dhcp_relay.py]: Replace with polling for status.

File: tests/generic_config_updater/test_dhcp_relay.py (hunk 4 above)
Comment (wen587) [reply tests/generic_config_updater/test_dhcp_relay.py]: Will ask see if there is the design

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+    logger.info("VLAN test ending ...")
+    logger.info("Delete VLAN intf")
+
+    cmds = []
+
+    for key, value in vlan_intfs_dict.items():
+        if not value['orig']:
+            cmds.append('config vlan member del {} {}'.format(key, rand_vlan_port))
+            cmds.append("config interface ip remove Vlan{} {}".format(key, value['ip'].upper()))
+            cmds.append('config vlan del {}'.format(key))
```
[the remark this answers is not in the document]
Comment (wen587) [reply tests/generic_config_updater/test_dhcp_relay.py:153]: Will change to 'config reload' if merge. The manual cleanup is for self test purpose.

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+def test_dhcp_relay_tc3_rm_nonexist(duthost, setup_vlan, init_dhcp_server_config, vlan_intfs_list):
+    """Test remove nonexisted dhcp server on default setup
+    """
+    if init_dhcp_server_config != "default_setup":
+        pytest.skip("Unsupported init config")
+
+    dhcp_rm_nonexist_json = [
+        {
+            "op": "remove",
+            "path": "/VLAN/Vlan"+ str(vlan_intfs_list[0]) + "/dhcp_servers/5"
```
[the remark this answers is not in the document]
Comment (wen587) [reply tests/generic_config_updater/test_dhcp_relay.py:208]: This is the default setup. So we only add 4 dhcp server for each. Any number large than 4(0 based) will be valid number

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+    
+    duthost.shell_cmds(cmds=cmds)
+    time.sleep(5) # This may take more time if orignal config setup has more dhcp servers under vlan
+    pytest_assert(duthost.is_service_fully_started('dhcp_relay'), "dhcp_relay service is not running during setup")
+    expect_res_success(duthost, expected_server_list, unexpected_server_list=[])
+
+def runout_setup():
+    pass
+
+#TODO: runout_setup
```
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply tests/generic_config_updater/test_dhcp_relay.py]: What does `runout_setup` mean?

File: tests/generic_config_updater/test_dhcp_relay.py (hunk 4 above)
Comment (qiluo-msft) [reply tests/generic_config_updater/test_dhcp_relay.py]: Low priority test case.

File: tests/generic_config_updater/gu_utils.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
+import json
+import time
```
Comment (qiluo-msft) [inline tests/generic_config_updater/gu_utils.py:2]: lgtm checker found it not used, and it's a true warning. Please respect lgtm alerts.

File: tests/common/fixtures/duthost_utils.py [hunk 11]
```diff
@@ -225,3 +226,79 @@ def vlan_ports_list(duthosts, rand_one_dut_hostname, rand_selected_dut, tbinfo,
...
 
     return vlan_ports_list
+
+def compare_network(src_ipprefix, dst_ipprefix):
+    src_network = ipaddress.IPv4Interface(src_ipprefix).network
+    dst_network = ipaddress.IPv4Interface(dst_ipprefix).network
+    return src_network.overlaps(dst_network)
+
+@pytest.fixture(scope="module")
+def utils_vlan_intfs_dict_orig(duthosts, rand_one_dut_hostname, tbinfo):
```
Comment (qiluo-msft) [inline tests/common/fixtures/duthost_utils.py:236]: >utils_vlan_intfs_dict_orig [](http://example.com/codeflow?start=4&length=26)

It's not easy to understand the function just from the name. They are implemented to be reused. Could you add some function level comments, explaining the purpose, parameter, and return values.

Comment (qiluo-msft) [reply tests/common/fixtures/duthost_utils.py:236]: Also applicable to other new functions in this file.

File: tests/generic_config_updater/gu_utils.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    duthost.file(path=tmpfile, state='absent')
+
+def apply_patch(duthost, json_data, dest_file):
+    duthost.copy(content=json.dumps(json_data, indent=4), dest=dest_file)
+
+    cmds = 'config apply-patch {}'.format(dest_file)
+
+    logger.info("Commands: {}".format(cmds))
+    output = duthost.shell(cmds, module_ignore_errors=True)
+    # output = duthost.shell(cmds)
```
Comment (qiluo-msft) [inline tests/generic_config_updater/gu_utils.py:22]: ># output = duthost.shell(cmds) [](http://example.com/codeflow?start=4&length=30)

Remove commented code.

File: tests/generic_config_updater/gu_utils.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        if "start-limit-hit" in line:
+            return True
+
+    return False
+
+def reset_start_limit_hit(duthost, container_name, threshold, interval, delay):
+    """Reset container if hit start-limit-hit
+    """
+    logger.info("Reset container '{}' due to start-limit-hit".format(container_name))
+    duthost.shell("sudo systemctl reset-failed {}.service".format(container_name))
```
Comment (qiluo-msft) [inline tests/generic_config_updater/gu_utils.py:71]: >duthost.shell [](http://example.com/codeflow?start=4&length=13)

Always check return code, in reality, we see systemctl fails.

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 14]
```diff
@@ -0,0 +1,450 @@
...
+    If added vlan_id is 108 and 109, it will add a dict as below
+    {108: {'ip': u'192.168.8.1/24', 'orig': False}, 109: {'ip': u'192.168.9.1/24', 'orig': False}}
+    '''
+    logger.info("vlan_intrfs_dict ORIG {}".format(utils_vlan_intfs_dict_orig))
+    vlan_intfs_dict = utils_vlan_intfs_dict_add(utils_vlan_intfs_dict_orig, 2)
+    logger.info("vlan_intrfs_dict FINAL {}".format(vlan_intfs_dict))
+    return vlan_intfs_dict
+
+@pytest.fixture(scope="module")
+def first_avai_vlan_port(rand_selected_dut, tbinfo):
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_dhcp_relay.py:38]: >first_avai_vlan_port [](http://example.com/codeflow?start=4&length=20)

avai -> available?

File: tests/generic_config_updater/test_dhcp_relay.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply tests/generic_config_updater/test_dhcp_relay.py:255]: Not to use a magic number? Let's assume in future we add 40 dhcp servers at the beginning. and 5 does not work.

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 15]
```diff
@@ -0,0 +1,450 @@
...
+    +-----------+------------------+-----------+----------------+-------------+-----------------------+
+    """
+    if init_dhcp_server_config != "default_setup":
+        pytest.skip("Unsupported init config")
+
+    dhcp_add_json = [
+        {
+            "op": "add",
+            "path": "/VLAN/Vlan" + str(vlan_intfs_list[1]) + "/dhcp_servers/4",
+            "value": "192.0." + str(vlan_intfs_list[1]) + ".5"
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_dhcp_relay.py:350]: >vlan_intfs_list[1] [](http://example.com/codeflow?start=36&length=18)

Is it possible that the value is too large and the IP is not valid? Check before use it this way?

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 16]
```diff
@@ -0,0 +1,450 @@
...
+    |           |                  |           |                |             | 192.0.109.4           |
+    +-----------+------------------+-----------+----------------+-------------+-----------------------+
+    """
+    if init_dhcp_server_config != "default_setup":
+        pytest.skip("Unsupported init config")
+
+    dhcp_rm_json = [
+        {
+        "op": "remove",
+        "path": "/VLAN/Vlan" + str(vlan_intfs_list[0]) + "/dhcp_servers/3"
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_dhcp_relay.py:310]: >3 [](http://example.com/codeflow?start=72&length=1)

also a magic number

File: tests/generic_config_updater/test_dhcp_relay.py [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
+
+    This function should generate two VLAN detail shown below
+    +-----------+------------------+-----------+----------------+-------------+-----------------------+
+    |   VLAN ID | IP Address       | Ports     | Port Tagging   | Proxy ARP   | DHCP Helper Address   |
+    +===========+==================+===========+================+=============+=======================+
+    |       108 | 192.168.8.1/24   | Ethernet4 | tagged         | disabled    |                       |
+    +-----------+------------------+-----------+----------------+-------------+-----------------------+
+    |       109 | 192.168.9.1/24   | Ethernet4 | tagged         | disabled    |                       |
+    +-----------+------------------+-----------+----------------+-------------+-----------------------+
+    '''
```
[the remark this answers is not in the document]
Comment (wen587) [reply tests/generic_config_updater/test_dhcp_relay.py:66]: Checked 2vlan setup. It is also `192.168.9.1/24` and `192.168.8.1/24`. These two doesnt overlap with 2vlan setup `192.168.0.1/22` and `192.168.4.1/22`

File: tests/generic_config_updater/test_dhcp_relay.py (hunk 14 above)
Comment (wen587) [reply tests/generic_config_updater/test_dhcp_relay.py:38]: Yes

File: tests/generic_config_updater/gu_utils.py (hunk 13 above)
Comment (wen587) [reply tests/generic_config_updater/gu_utils.py]: Added

File: tests/generic_config_updater/gu_utils.py (hunk 12 above)
Comment (wen587) [reply tests/generic_config_updater/gu_utils.py]: removed

File: tests/common/fixtures/duthost_utils.py (hunk 11 above)
Comment (wen587) [reply tests/common/fixtures/duthost_utils.py:236]: Added

File: tests/generic_config_updater/test_dhcp_relay.py (hunk 15 above)
Comment (wen587) [reply tests/generic_config_updater/test_dhcp_relay.py:368]: No. It is restricted by `utils_vlan_intfs_dict_add` which in range(0, 255)

File: tests/generic_config_updater/test_dhcp_relay.py (hunk 16 above)
Comment (wen587) [reply tests/generic_config_updater/test_dhcp_relay.py:325]: Here I use magic number due that if the index change, the json will look much different. Here I choose to remove the last element for simplicity but also check the correctness. If I choose to remove other index such as 1. The json_patch will look like this:
```
[
 {
  "op": "remove",
  "path": "/VLAN/Vlan108/dhcp_servers/3"
 },
 {
  "op": "replace",
  "path": "/VLAN/Vlan108/dhcp_servers/2",
  "value": "192.0.108.4"
 },
 {
  "op": "replace",
  "path": "/VLAN/Vlan108/dhcp_servers/1",
  "value": "192.0.108.3"
 }
]
```
The vlan dhcp server is not simple `key->value` pair. It looks like a `key->indexed value`

**Review by qiluo-msft — APPROVED** [review]
